pallets/deitos/src/lib.rs

```diff
@@ -113,54 +135,180 @@ pub mod pallet {
     #[pallet::generate_deposit(pub(super) fn deposit_event)]
     pub enum Event<T: Config> {
         /// A user has successfully set a new value.
-        SomethingStored {
-            /// The new value set.
-            something: u32,
-            /// The account who set the new value.
-            who: T::AccountId,
+        IPRegistered {
+            ip: T::AccountId,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        },
+        IPStorageUpdated {
+            ip: T::AccountId,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        },
+        IPStatusChanged {
+            ip: T::AccountId,
+            status: IPStatus,
+        },
+        IPShutdown {
+            ip: T::AccountId,
         },
     }
 
     /// information.
     #[pallet::error]
     pub enum Error<T> {
         /// The value retrieved was `None` as no value was previously set.
-        NoneValue,
-        StorageOverflow,
         NonExistentStorageValue,
+        /// Math overflow
+        Overflow,
+        /// Insufficient storage
+        InsufficientStorage,
+        /// On going agreements
+        OnGoingAgreements,
     }
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         #[pallet::call_index(0)]
         #[pallet::weight(T::WeightInfo::register_ip())]
         /* to add more parameters*/
-        pub fn register_ip(origin: OriginFor<T>, price_unit: BalanceOf<T>) -> DispatchResult {
+        pub fn register_ip(
+            origin: OriginFor<T>,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        ) -> DispatchResult {
             // Check that the extrinsic was signed and get the signer.
-            let who = ensure_signed(origin)?;
+            let ip = ensure_signed(origin)?;
+            let bounding_amount = IPBoundingAmount::<T>::get()?;
+
+            T::Currency::hold(&HoldReason::Transfer.into(), &ip, bounding_amount)?;
```

Comment:
i'm not sure if the `RuntimeHoldReason` supports nested enums, but we might try ... worst case i can switch it to `IPInitialDeposit`

Reply:
AFAIK you can define your hold reasons in the pallet like this:
```
#[pallet::composite_enum]
pub enum HoldReason {
    #[codec(index = 0)]
    Transfer,
}
```

when runtime is constructed all enums across all pallets marked with `pallet::composite_enum` are combined into a single enum which becomes `RuntimeHoldReason`

Reply:
Please confirm if what i understood is that you want to have nested enums like:

`HoldReason::Deposit::IPInitialDeposit`

or you were just proposing diffrent options for the `HoldReason` composite enum.

Reply:
no-no, I am suggesting just a different naming here:
```
#[pallet::composite_enum]
pub enum HoldReason {
    #[codec(index = 0)]
    Deposit,
}
```
because the reason why we are holding this is a deposit, not a transport. But I am not sure if just `Deposit` is enough. Might be better to call it `InitialDeposit` or even smth else. 

There also be some other types of hold reason, like consumer's deposit to secure the payment.